f3/certstore.go

```diff
@@ -0,0 +1,189 @@
+}
+
+var ErrCertAlreadyExists = errors.New("certificate already exists in the CertStore")
+var ErrCertNotFound = errors.New("certificate not found")
+
+// CertStore is responsible for storying and relaying information about new finality certificates
+type CertStore struct {
+	dsLk     sync.Mutex
+	ds       datastore.Datastore
+	busCerts broadcast.Channel[Cert]
+}
+
+func NewCertStore(ctx context.Context, ds datastore.Datastore) (*CertStore, error) {
+	cs := &CertStore{
+		ds: namespace.Wrap(ds, datastore.NewKey("/certstore")),
+	}
+	latestCert, err := cs.loadLatest(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("loading latest cert: %w", err)
+	}
+	if latestCert != nil {
+		cs.busCerts.Publish(*latestCert)
+	}
+
+	return cs, nil
+}
+
+var certStoreLatestKey = datastore.NewKey("/latestCert")
+
+func (cs *CertStore) loadLatest(ctx context.Context) (*Cert, error) {
+	cb, err := cs.ds.Get(ctx, certStoreLatestKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("getting latest cert: %w", err)
+	}
+	var c Cert
+	err = c.UnmarshalBinary(cb)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshalling latest cert: %w", err)
+	}
+	return &c, err
+}
+
+func (cs *CertStore) Latest() *Cert {
+	return cs.busCerts.Last()
+}
+
+func (cs *CertStore) Get(ctx context.Context, instance uint64) (*Cert, error) {
+	cs.dsLk.Lock()
+	b, err := cs.ds.Get(ctx, cs.keyForInstance(instance))
+	cs.dsLk.Unlock()
+
+	if errors.Is(err, datastore.ErrNotFound) {
+		return nil, xerrors.Errorf("cert at %d: %w", instance, ErrCertNotFound)
+	}
+	if err != nil {
+		return nil, xerrors.Errorf("accessing cert in datastore: %w", err)
+	}
+
+	var c Cert
+	if err := c.UnmarshalBinary(b); err != nil {
+		return nil, xerrors.Errorf("unmarshalling cert: %w", err)
+	}
+	return &c, err
+}
+
+// GetRange returns a range of certs from start to end inclusive.
+// If it encouters missing cert, it returns a wrapped ErrCertNotFound and the available certs
+func (cs *CertStore) GetRange(ctx context.Context, start uint64, end uint64) ([]Cert, error) {
+	if start > end {
+		return nil, xerrors.Errorf("start is larger then end: %d > %d", start, end)
+	}
+	bCerts := make([][]byte, 0, start-end+1)
+	i := start
+
+	cs.dsLk.Lock()
+	for ; i <= end; i++ {
+		b, err := cs.ds.Get(ctx, cs.keyForInstance(i))
+		if errors.Is(err, datastore.ErrNotFound) {
+			break
+		}
+		if err != nil {
+			cs.dsLk.Unlock()
+			return nil, xerrors.Errorf("accessing cert at %d for range request: %w", i, err)
+		}
+
+		bCerts = append(bCerts, b)
+	}
+	cs.dsLk.Unlock()
+
+	certs := make([]Cert, 0, len(bCerts))
+	for i := 0; i < len(bCerts); i++ {
+		err := certs[i].UnmarshalBinary(bCerts[i])
+		if err != nil {
+			return nil, xerrors.Errorf("unmarshalling a cert at i=%d, instance %d: %w", i, start+uint64(i), err)
+		}
+	}
+
+	if uint64(len(certs)) < start-end {
+		return certs, xerrors.Errorf("cert at %d: %w", i, ErrCertNotFound)
```

Comment:
What enforces that this range is fully populated? Should the store enforce that? If yes, how can the user ensure they use the API correctly? If no, how can the user understand the gaps?

Reply:
We should enforce it, I will add it.